lib/HashbangPlugin.js

```diff
@@ -0,0 +1,59 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author @wh0
+*/
+
+"use strict";
+
+const ConstDependency = require("./dependencies/ConstDependency");
+
+/** @typedef {import("./Compiler")} Compiler */
+
+class HashbangPlugin {
+	/**
+	 * Apply the plugin
+	 * @param {Compiler} compiler the compiler instance
+	 * @returns {void}
+	 */
+	apply(compiler) {
+		compiler.hooks.compilation.tap(
+			"HashbangPlugin",
+			(compilation, { normalModuleFactory }) => {
+				const handler = parser => {
+					parser.hooks.program.tap("HashbangPlugin", (ast, comments) => {
+						const firstComment = comments[0];
+						if (
+							firstComment &&
+							parser.state.module
+								.originalSource()
+								.source()
+								.slice(
+									firstComment.start,
+									Math.min(firstComment.start + 2, firstComment.end)
+								) === "#!"
```

Comment:
Why we need get source? Why do not use `firstComment`?

Reply:
it only has the "value," which leaves out the comment characters. e.g. here it would be `/use/bin/env node`.

Reply:
I think we should improve it on parser side :thinking: otherwise it reduce perf

Reply:
that might be useful to improve. but could you describe the perf impact of getting the source here? my expectation is that originalSource is a RawSource that already had the file in memory from when we parsed it in the first place

Reply:
If you have a lot of files (very many) if will read content of each, but we need run this hook only where we found shebang

Reply:
can you give an example of a code path where this would _read_ the content? and/or correct the below

here's the code path I was looking at, where it would not:

1. normal module calls `source()` to give input to the module's parser https://github.com/webpack/webpack/blob/2fcb50cf0fdca6651ffbc15f79b984b6f07f4651/lib/NormalModule.js#L844
2. raw source either accesses already-in-memory source https://github.com/webpack/webpack-sources/blob/master/lib/RawSource.js#L31
3. javascript parser calls `program` hook https://github.com/webpack/webpack/blob/2fcb50cf0fdca6651ffbc15f79b984b6f07f4651/lib/javascript/JavascriptParser.js#L3271
4. we get here, where `source()` again accesses already-in-memory source like in step 2

Reply:
What about create new hook (`sheband`)?

Reply:
I started doing something like that, but then it started to look like we could still do the actual detection of whether a comment is a shebang within HashbangPlugin, if we would just pass the source that we give to the parser to the `program` hook as well. I've tentatively added that to this branch. This should make it clearer that we're not doing any extra reading in order to inspect the source, but again, I'm not convinced that the previous approach did any extra reading. My thoughts on this, and it'd be good to hear yours as well:

Pros:
- shebang detection takes place where it ought to, in HashbangPlugin
- potentially more useful in general in case other plugins also want to use the source

Cons:
- slightly more documentation work to add the new `program` hook parameter
- less theoretically pure and potentially confusing with this additional way to get the source from withing the `program` hook, next to the old method of accessing `module.originalSource().source()`. especially that `parsedSource` is the string form

Neutral:
- I'm not convinced this saves on resources. Type system suggests previous version might have been incorrect with `module.originalSource().source()` might have been `Buffer`, which we would have had to `toString` which would be expensive, but I don't have any evidence that this could have actually happened for a Javascript module.
- if later plugins actually wanted a higher level of abstraction to tap a shebang hook, this doesn't do that, but it doesn't make it harder either. HashbangPlugin could be in charge of that hook.

Reply:
/cc @sokra need your advice